spec/lib/search_logic_spec.rb

```diff
@@ -0,0 +1,98 @@
+RSpec.describe SearchLogic do
+    it "saves new attractions" do
+        geocode_stub = stub_request(:get, %r{https://api.geoapify.com/v1/geocode/search\?apiKey=[^&]+&format=json&text=Venice,%20Italy}).to_return(status: 200, body: File.read("spec/responses/geocode_venice.json"))
+        
+        places_stub = stub_request(:get, venice_request_url("tourism.attraction")).to_return(status: 200, body: File.read("spec/responses/attractions_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "attractions")
+            expect(result).to all(be_a(Attraction))
+        }.to change(City, :count).by(1).and change(Attraction, :count).by(20)
+
+        expect(geocode_stub).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+
+    it "saves new hotels" do
+        geocode_venice
+
+        places_stub = stub_request(:get, venice_request_url("accommodation.hotel")).to_return(status: 200, body: File.read("spec/responses/hotels_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "hotels")
+            expect(result).to all(be_a(Hotel))
+        }.to change(City, :count).by(1).and change(Hotel, :count).by(20)
+
+        expect(geocode_venice).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+    
+    it "saves new restaurants" do
+        geocode_venice
+
+        places_stub = stub_request(:get, venice_request_url("catering.restaurant")).to_return(status: 200, body: File.read("spec/responses/restaurants_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "restaurants")
+            expect(result).to all(be_a(Restaurant))
+        }.to change(City, :count).by(1).and change(Restaurant, :count).by(20)
+
+        expect(geocode_venice).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+
+    it "does not create a city if it is already in the database" do
+        places_stub = stub_request(:get, venice_request_url("tourism.attraction")).to_return(status: 200, body: File.read("spec/responses/attractions_venice.json"))
+
+        city = create_venice
+
+        expect {
+            SearchLogic.get_places_from_city_string(city.name, "attractions")
+        }.to change(City, :count).by(0)
+
+        expect(geocode_venice).to have_been_requested.times(0)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+
+    it "does not create a place that is already in the database" do
```

Comment:
Since you have the `if places.count == 20` check now, there are two different behaviours when you have places that are already in the database:

1. There are not exactly 20 places, so it makes another call to Geoapify. This scenario covers that.
2. There are exactly 20 places, so no calls are made to Geoapify. You'll want a scenario for this.